src/gesv_rbt.cc

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) 2020-2022, University of Tennessee. All rights reserved.
+// SPDX-License-Identifier: BSD-3-Clause
+// This program is free software: you can redistribute it and/or modify it under
+// the terms of the BSD 3-Clause license. See the accompanying LICENSE file.
+
+#include "slate/slate.hh"
+#include "internal/internal.hh"
+#include "internal/internal_util.hh"
+
+namespace slate {
+
+//------------------------------------------------------------------------------
+/// Distributed parallel LU factorization and solve.
+///
+/// Computes the solution to a system of linear equations
+/// \[
+///     A X = B,
+/// \]
+/// where $A$ is an n-by-n matrix and $X$ and $B$ are n-by-nrhs matrices.
+///
+/// gesv_rbt first transforms the matrix with random butterfly transforms,
+/// factorizes the transformed matrix using getrf_nopiv, and uses this
+/// factorization within an iterative refinement procedure to produce a
+/// solution with full normwise backward error quality (see below).  If the
+/// approach fails and the UseFallbackSolver is true, the problem is re-solved
+/// with a partial pivoted factorization.
+///
+/// The iterative refinement process is stopped if iter > itermax or
+/// for all the RHS, $1 \le j \le nrhs$, we have:
+///     $\norm{r_j}_{inf} < tol \norm{x_j}_{inf} \norm{A}_{inf},$
+/// where:
+/// - iter is the number of the current iteration in the iterative refinement
+///    process
+/// - $\norm{r_j}_{inf}$ is the infinity-norm of the residual, $r_j = Ax_j - b_j$
+/// - $\norm{x_j}_{inf}$ is the infinity-norm of the solution
+/// - $\norm{A}_{inf}$ is the infinity-operator-norm of the matrix $A$
+///
+//------------------------------------------------------------------------------
+/// @tparam scalar_t
+///     One of float, double, std::complex<float>, std::complex<double>.
+//------------------------------------------------------------------------------
+/// @param[in,out] A
+///     On entry, the n-by-n matrix $A$ to be factored.
+///     On exit, the factors $L$ and $U$ from the factorization $A = P L U$;
+///     the unit diagonal elements of $L$ are not stored.
+///
+/// @param[out] pivots
+///     The pivot indices that define the permutation matrix $P$.
+///
+/// @param[in] B
+///     On entry, the n-by-nrhs right hand side matrix $B$.
+///
+/// @param[out] X
+///     On exit, if return value = 0, the n-by-nrhs solution matrix $X$.
+///
+/// @param[out] iter
+///     The number of the iterations in the iterative refinement
+///     process, needed for the convergence. If iterative refinement failed,
+///     it is set to -(1+itermax), regardless of whether the fallback solver
+///     was used.
+///
+/// @param[in] opts
+///     Additional options, as map of name = value pairs. Possible options:
+///     - Option::Lookahead:
+///       Number of panels to overlap with matrix updates.
+///       lookahead >= 0. Default 1.
```

Comment:
I think the default value of the `lookahead` should be defined in this function, right?

Reply:
I rely on the default in `getrf_nopiv`, similar to regular `gesv` and `gesv_mixed`.